File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/SignatureScheme.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bcrand.RandomNumberGenerator
import com.blockchaincommons.bcrand.SecureRandomNumberGenerator

/**
 * Supported digital signature schemes.
 *
 * This enum represents the various signature schemes supported in this
 * library, including elliptic curve schemes (ECDSA, Schnorr), Edwards
 * curve schemes (Ed25519), post-quantum schemes (ML-DSA), and
 * SSH-specific algorithms.
 */
enum class SignatureScheme {
    /** BIP-340 Schnorr signature scheme, used in Bitcoin Taproot (default). */
    Schnorr,

    /** ECDSA signature scheme using the secp256k1 curve. */
    ECDSA,

    /** Ed25519 signature scheme (RFC 8032). */
    Ed25519,

    /** ML-DSA44 post-quantum signature scheme (NIST level 2). */
    MLDSA44,

    /** ML-DSA65 post-quantum signature scheme (NIST level 3). */
    MLDSA65,

    /** ML-DSA87 post-quantum signature scheme (NIST level 5). */
    MLDSA87,

    /** Ed25519 signature scheme for SSH. */
    SshEd25519,

    /** DSA signature scheme for SSH. */
    SshDsa,

    /** ECDSA signature scheme with NIST P-256 curve for SSH. */
    SshEcdsaP256,

    /** ECDSA signature scheme with NIST P-384 curve for SSH. */
    SshEcdsaP384;

    /**
     * Creates a new key pair for this signature scheme using the system's
     * secure random number generator.
     *
     * @return a pair of signing private key and its corresponding public key
     * @throws UnsupportedOperationException for SSH variants that are not
     *   yet fully translated
     */
    fun keypair(): Pair<SigningPrivateKey, SigningPublicKey> {
        return when (this) {
            MLDSA44 -> {
                val (privKey, pubKey) = MLDSA.MLDSA44.keypair()
                SigningPrivateKey.MLDSAKey(privKey) to SigningPublicKey.MLDSAKey(pubKey)
            }
            MLDSA65 -> {
                val (privKey, pubKey) = MLDSA.MLDSA65.keypair()
                SigningPrivateKey.MLDSAKey(privKey) to SigningPublicKey.MLDSAKey(pubKey)
            }
            MLDSA87 -> {
                val (privKey, pubKey) = MLDSA.MLDSA87.keypair()
                SigningPrivateKey.MLDSAKey(privKey) to SigningPublicKey.MLDSAKey(pubKey)
            }
            else -> keypairUsing(SecureRandomNumberGenerator(), "")
        }
    }

    /**
     * Creates a key pair for this signature scheme using a provided random
     * number generator with an optional comment.
     *
     * The comment is only used for SSH keys and is ignored for other schemes.
     *
     * @param rng the random number generator to use
     * @param comment an optional string comment for SSH keys
     * @return a pair of signing private key and its corresponding public key
     * @throws UnsupportedOperationException for SSH variants and ML-DSA
     *   variants (which do not support deterministic generation with a
     *   custom RNG)
     */
    fun keypairUsing(
        rng: RandomNumberGenerator,
        comment: String = "",
    ): Pair<SigningPrivateKey, SigningPublicKey> {
        return when (this) {
            Schnorr -> {
                val ecKey = ECPrivateKey.createUsing(rng)
                val privateKey = SigningPrivateKey.SchnorrKey(ecKey)
                val publicKey = privateKey.publicKey()
                privateKey to publicKey
            }
            ECDSA -> {
                val ecKey = ECPrivateKey.createUsing(rng)
                val privateKey = SigningPrivateKey.ECDSAKey(ecKey)
                val publicKey = privateKey.publicKey()
                privateKey to publicKey
            }
            Ed25519 -> {
                val edKey = Ed25519PrivateKey.createUsing(rng)
                val privateKey = SigningPrivateKey.Ed25519Key(edKey)
                val publicKey = privateKey.publicKey()
                privateKey to publicKey
            }
            MLDSA44, MLDSA65, MLDSA87 -> {
                throw UnsupportedOperationException(
                    "Deterministic keypair generation not supported for ML-DSA"
                )
            }
            SshEd25519, SshDsa, SshEcdsaP256, SshEcdsaP384 -> {
                throw UnsupportedOperationException(
                    "SSH key generation not yet supported"
                )
            }
        }
    }
}

File: csharp/BCComponents/BCComponents/SignatureScheme.cs
using BlockchainCommons.BCRand;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// Supported digital signature schemes.
/// </summary>
/// <remarks>
/// This enum represents the various signature schemes supported, including
/// elliptic curve schemes (ECDSA, Schnorr), Edwards curve schemes (Ed25519),
/// post-quantum schemes (ML-DSA), and SSH-specific algorithms.
/// </remarks>
public enum SignatureScheme
{
    /// <summary>BIP-340 Schnorr signature scheme, used in Bitcoin Taproot (default).</summary>
    Schnorr,

    /// <summary>ECDSA signature scheme using the secp256k1 curve.</summary>
    Ecdsa,

    /// <summary>Ed25519 signature scheme (RFC 8032).</summary>
    Ed25519,

    /// <summary>ML-DSA44 post-quantum signature scheme (NIST level 2).</summary>
    MLDSA44,

    /// <summary>ML-DSA65 post-quantum signature scheme (NIST level 3).</summary>
    MLDSA65,

    /// <summary>ML-DSA87 post-quantum signature scheme (NIST level 5).</summary>
    MLDSA87,

    /// <summary>Ed25519 signature scheme for SSH.</summary>
    SshEd25519,

    /// <summary>DSA signature scheme for SSH.</summary>
    SshDsa,

    /// <summary>ECDSA signature scheme with NIST P-256 curve for SSH.</summary>
    SshEcdsaP256,

    /// <summary>ECDSA signature scheme with NIST P-384 curve for SSH.</summary>
    SshEcdsaP384,
}

/// <summary>
/// Extension methods for <see cref="SignatureScheme"/>.
/// </summary>
public static class SignatureSchemeExtensions
{
    /// <summary>
    /// Creates a new key pair for the signature scheme using the system's
    /// secure random number generator.
    /// </summary>
    /// <param name="scheme">The signature scheme to generate keys for.</param>
    /// <returns>A tuple containing a signing private key and its corresponding public key.</returns>
    public static (SigningPrivateKey PrivateKey, SigningPublicKey PublicKey) Keypair(
        this SignatureScheme scheme)
    {
        return scheme.KeypairOpt("");
    }

    /// <summary>
    /// Creates a new key pair for the signature scheme with an optional comment.
    /// </summary>
    /// <remarks>
    /// The comment is only used for SSH keys and is ignored for other schemes.
    /// </remarks>
    /// <param name="scheme">The signature scheme to generate keys for.</param>
    /// <param name="comment">A string comment to include with SSH keys.</param>
    /// <returns>A tuple containing a signing private key and its corresponding public key.</returns>
    public static (SigningPrivateKey PrivateKey, SigningPublicKey PublicKey) KeypairOpt(
        this SignatureScheme scheme,
        string comment)
    {
        switch (scheme)
        {
            case SignatureScheme.Schnorr:
            {
                var privateKey = SigningPrivateKey.NewSchnorr(ECPrivateKey.New());
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.Ecdsa:
            {
                var privateKey = SigningPrivateKey.NewEcdsa(ECPrivateKey.New());
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.Ed25519:
            {
                var privateKey = SigningPrivateKey.NewEd25519(Ed25519PrivateKey.New());
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.MLDSA44:
            {
                var (privKey, pubKey) = MLDSALevel.MLDSA44.Keypair();
                return (SigningPrivateKey.NewMldsa(privKey), SigningPublicKey.FromMldsa(pubKey));
            }
            case SignatureScheme.MLDSA65:
            {
                var (privKey, pubKey) = MLDSALevel.MLDSA65.Keypair();
                return (SigningPrivateKey.NewMldsa(privKey), SigningPublicKey.FromMldsa(pubKey));
            }
            case SignatureScheme.MLDSA87:
            {
                var (privKey, pubKey) = MLDSALevel.MLDSA87.Keypair();
                return (SigningPrivateKey.NewMldsa(privKey), SigningPublicKey.FromMldsa(pubKey));
            }
            case SignatureScheme.SshEd25519:
            {
                var privateKey = SshKeyHelper.GenerateSshSigningPrivateKey(SshKeyHelper.SshAlgorithm.Ed25519, comment);
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.SshDsa:
            {
                var privateKey = SshKeyHelper.GenerateSshSigningPrivateKey(SshKeyHelper.SshAlgorithm.Dsa, comment);
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.SshEcdsaP256:
            {
                var privateKey = SshKeyHelper.GenerateSshSigningPrivateKey(SshKeyHelper.SshAlgorithm.EcdsaP256, comment);
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.SshEcdsaP384:
            {
                var privateKey = SshKeyHelper.GenerateSshSigningPrivateKey(SshKeyHelper.SshAlgorithm.EcdsaP384, comment);
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(scheme), scheme, "Unsupported signature scheme");
        }
    }

    /// <summary>
    /// Creates a key pair for the signature scheme using a provided random number generator.
    /// </summary>
    /// <remarks>
    /// This allows for deterministic key generation when using a seeded RNG.
    /// Note that not all signature schemes support deterministic generation (e.g., ML-DSA does not).
    /// </remarks>
    /// <param name="scheme">The signature scheme to generate keys for.</param>
    /// <param name="rng">A random number generator to use.</param>
    /// <param name="comment">A string comment to include with SSH keys.</param>
    /// <returns>A tuple containing a signing private key and its corresponding public key.</returns>
    /// <exception cref="BCComponentsException">
    /// Thrown if the scheme does not support deterministic generation.
    /// </exception>
    public static (SigningPrivateKey PrivateKey, SigningPublicKey PublicKey) KeypairUsing(
        this SignatureScheme scheme,
        IRandomNumberGenerator rng,
        string comment)
    {
        switch (scheme)
        {
            case SignatureScheme.Schnorr:
            {
                var privateKey = SigningPrivateKey.NewSchnorr(ECPrivateKey.NewUsing(rng));
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.Ecdsa:
            {
                var privateKey = SigningPrivateKey.NewEcdsa(ECPrivateKey.NewUsing(rng));
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            case SignatureScheme.Ed25519:
            {
                var privateKey = SigningPrivateKey.NewEd25519(Ed25519PrivateKey.NewUsing(rng));
                var publicKey = privateKey.PublicKey();
                return (privateKey, publicKey);
            }
            default:
                throw BCComponentsException.General(
                    "Deterministic keypair generation not supported for this signature scheme");
        }
    }
}
